src/python/pants/jvm/compile.py

```diff
@@ -47,6 +54,34 @@ class _ClasspathEntryRequestClassification(Enum):
     INCOMPATIBLE = auto()
 
 
+@dataclass(frozen=True)
+class JVMRequestTypes:
+    classpath_entry_requests: tuple[type[ClasspathEntryRequest], ...]
+    code_generator_requests: FrozenDict[type[SourcesField], type[ClasspathEntryRequest]]
+
+
+@rule
+def calculate_jvm_request_types(union_membership: UnionMembership) -> JVMRequestTypes:
+    cpe_impls = union_membership.get(ClasspathEntryRequest)
+    b: dict[type[Field], type[ClasspathEntryRequest]] = {}
+    for impl in cpe_impls:
+        for field_set in impl.field_sets:
+            for field in field_set.required_fields:
+                # Assume only one impl per field (normally sound)
+                b[field] = impl
```

Comment:
A field set could require both a source field and a JDK field, for example... I expect that this assumption only holds for source fields. So you could maybe filter to source fields? But even then, something might have multiple source fields I think? cc @Eric-Arellano

Reply:
Having multiple source fields is almost guaranteed to not work elsewhere. This is because we use `tgt.get(SourceField)` in dozens of places. I wouldn't worry about multiple sources.

Fwit, we have `SecondaryOwnerMixin` to allow you to get some sources-like behavior without worrying about only one field. This is how `pex_binary` works with file arguments even though it doesn't actually have sources.

Reply:
The concrete `ClasspathEntryRequest` classes aren't dependent on JVM version though?

Reply:
Re-reading: Filtering to source fields makes sense for clarity (and I might make that change for that reason), but is unlikely to impact functionality: Given that code generators' outputs are source fields, only the source fields are going to be matched in the end.

Reply:
added a comment to this effect.